holoviews/element/graphs.py

```diff
@@ -180,12 +178,31 @@ def range(self, dimension, data_range=True):
 
     def dimensions(self, selection='all', label=False):
         dimensions = super(Graph, self).dimensions(selection, label)
-        if self.nodes and selection == 'ranges':
-            return dimensions+self.nodes.dimensions(selection, label)
+        if selection == 'ranges':
+            if self._nodes:
+                node_dims = self.nodes.dimensions(selection, label)
+            else:
+                node_dims = Nodes.kdims+Nodes.vdims
+                if label in ['name', True, 'short']:
+                    node_dims = [d.name for d in node_dims]
+                elif label in ['long', 'label']:
+                    node_dims = [d.label for d in node_dims]
+            return dimensions+node_dims
         return dimensions
 
 
     @property
+    def nodes(self):
+        """
+        Computes the node positions the first time they are requested
+        if no explicit node information was supplied.
+        """
```

Comment:
I still think this is a bit weird. I won't revert it though but I would like to keep the option to reconsider later...